### PR TITLE
[gcs/ha] Fix cpp tests related to redis removal

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

This PR fixed cpp tests and also make ray cpp able to pass.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -165,6 +165,12 @@
     - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/travis/upload_build_info.sh; fi }; trap cleanup EXIT
     - ./ci/travis/ci.sh test_cpp
 
+- label: ":redis: Worker"
```
Comment (mwtian) [inline .buildkite/pipeline.yml:168]: C++ Worker?

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -165,6 +165,12 @@
...
+  conditions: [ "RAY_CI_CPP_AFFECTED" ]
+  commands:
+    - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/travis/upload_build_info.sh; fi }; trap cleanup EXIT
+    - RAY_bootstrap_with_gcs=1 RAY_gcs_grpc_based_pubsub= RAY_gcs_storage=memory ./ci/travis/ci.sh test_cpp
```
Comment (mwtian) [inline .buildkite/pipeline.yml:172]: `RAY_gcs_grpc_based_pubsub=1`?

File: cpp/src/ray/config_internal.cc [hunk 3]
```diff
@@ -118,12 +118,14 @@ void ConfigInternal::Init(RayConfig &config, int argc, char **argv) {
     startup_token = absl::GetFlag<int64_t>(FLAGS_startup_token);
   }
   if (worker_type == WorkerType::DRIVER && run_mode == RunMode::CLUSTER) {
-    if (redis_ip.empty()) {
+    if ((::RayConfig::instance().bootstrap_with_gcs() && gcs_ip.empty()) ||
```
Comment (mwtian) [inline cpp/src/ray/config_internal.cc:121]: Is the leading `::` necessary?

File: cpp/src/ray/config_internal.cc [hunk 4]
```diff
@@ -118,12 +118,14 @@ void ConfigInternal::Init(RayConfig &config, int argc, char **argv) {
...
+        redis_ip.empty()) {
       auto ray_address_env = std::getenv("RAY_ADDRESS");
       if (ray_address_env) {
         RAY_LOG(DEBUG) << "Initialize Ray cluster address to \"" << ray_address_env
                        << "\" from environment variable \"RAY_ADDRESS\".";
-        SetRedisAddress(ray_address_env);
+        RAY_LOG(ERROR) << "??? " << config.address;
```
Comment (mwtian) [inline cpp/src/ray/config_internal.cc:127]: Remove?

File: cpp/src/ray/config_internal.cc [hunk 5]
```diff
@@ -145,11 +147,16 @@ void ConfigInternal::Init(RayConfig &config, int argc, char **argv) {
   }
 };
 
-void ConfigInternal::SetRedisAddress(const std::string address) {
+void ConfigInternal::SetBootstrapAddress(const std::string &address) {
```
Comment (mwtian) [inline cpp/src/ray/config_internal.cc:150]: nit: maybe accept `std::string_view` instead, so `substr()` avoids copying.

File: cpp/src/ray/config_internal.cc [hunk 6]
```diff
@@ -145,11 +147,16 @@ void ConfigInternal::Init(RayConfig &config, int argc, char **argv) {
...
   auto pos = address.find(':');
   RAY_CHECK(pos != std::string::npos);
-  redis_ip = address.substr(0, pos);
-  redis_port = std::stoi(address.substr(pos + 1, address.length()));
+  if (::RayConfig::instance().bootstrap_with_gcs()) {
+    gcs_ip = address.substr(0, pos);
+    redis_port = std::stoi(address.substr(pos + 1, address.length()));
```
Comment (mwtian) [inline cpp/src/ray/config_internal.cc:155]: This should be `gcs_port`? Or maybe we can use `port` or `bootstrap_port` for both `redis_port` and `gcs_port`?

File: src/ray/gcs/gcs_client/test/gcs_client_test.cc [hunk 7]
```diff
@@ -28,35 +28,53 @@ namespace ray {
...
     config_.node_ip_address = "127.0.0.1";
     config_.enable_sharding_conn = false;
-    config_.redis_port = TEST_REDIS_SERVER_PORTS.front();
+
     // Tests legacy code paths. The poller and broadcaster have their own dedicated unit
     // test targets.
     config_.grpc_based_resource_broadcast = false;
-    config_.grpc_pubsub_enabled = use_gcs_pubsub_;
+    config_.grpc_pubsub_enabled = enable_gcs_bootstrap_;
+    ;
```
Comment (mwtian) [inline src/ray/gcs/gcs_client/test/gcs_client_test.cc:77]: Remove?

File: src/ray/gcs/gcs_client/test/gcs_client_test.cc [hunk 8]
```diff
@@ -80,9 +98,14 @@ class GcsClientTest : public ::testing::TestWithParam<bool> {
...
-                                  config_.redis_password);
-    gcs_client_.reset(new gcs::GcsClient(options));
+    if (enable_gcs_bootstrap_) {
+      gcs::GcsClientOptions options("127.0.0.1:5397");
+      gcs_client_.reset(new gcs::GcsClient(options));
+    } else {
+      gcs::GcsClientOptions options(config_.redis_address, config_.redis_port,
+                                    config_.redis_password);
+      gcs_client_.reset(new gcs::GcsClient(options));
+    }
```
Comment (mwtian) [inline src/ray/gcs/gcs_client/test/gcs_client_test.cc:108]:
```suggestion
    if (enable_gcs_bootstrap_) {
      gcs::GcsClientOptions options("127.0.0.1:5397");
      gcs_client_ = std::make_unique<gcs::GcsClient>(options);
    } else {
      gcs::GcsClientOptions options(config_.redis_address, config_.redis_port,
                                    config_.redis_password);
      gcs_client_ = std::make_unique<gcs::GcsClient>(options);
    }
```

File: src/ray/gcs/gcs_client/test/gcs_client_test.cc [hunk 9]
```diff
@@ -878,6 +903,10 @@ TEST_P(GcsClientTest, TestWorkerTableResubscribe) {
 }
 
 TEST_P(GcsClientTest, TestGcsTableReload) {
+  // Restart gcs only work with redis.
+  if (RayConfig::instance().bootstrap_with_gcs()) {
```
Comment (mwtian) [inline src/ray/gcs/gcs_client/test/gcs_client_test.cc:907]: What will it take to set up external Redis for this test, in a future PR?

File: src/ray/gcs/gcs_client/gcs_client.h [hunk 10]
```diff
@@ -53,7 +53,10 @@ class GcsClientOptions {
         password_(password),
         enable_sync_conn_(enable_sync_conn),
         enable_async_conn_(enable_async_conn),
-        enable_subscribe_conn_(enable_subscribe_conn) {}
+        enable_subscribe_conn_(enable_subscribe_conn) {
+    RAY_LOG(DEBUG) << "Connect to gcs server via redis: " << redis_ip << ":"
+                   << redis_port;
+  }
```
Comment (MissiontoMars) [inline src/ray/gcs/gcs_client/gcs_client.h:59]: Should be removed?

File: src/ray/gcs/gcs_client/gcs_client.h [hunk 11]
```diff
@@ -63,6 +66,7 @@ class GcsClientOptions {
     RAY_CHECK(address.size() == 2);
     gcs_address_ = address[0];
     gcs_port_ = std::stoi(address[1]);
+    RAY_LOG(DEBUG) << "Connect to gcs server via address: " << gcs_address;
```
Comment (MissiontoMars) [inline src/ray/gcs/gcs_client/gcs_client.h:69]: Should be removed?

File: src/ray/gcs/gcs_client/gcs_client.h (hunk 10 above)
Comment (iycheng) [reply src/ray/gcs/gcs_client/gcs_client.h:59]: It's debugging log. It's useful for debugging purposes. So I'd prefer leaving here.

File: cpp/src/ray/config_internal.cc (hunk 3 above)
Comment (iycheng) [reply cpp/src/ray/config_internal.cc:121]: Yes. In ray cpp, there is another RayConfig unfortunately.

File: src/ray/gcs/gcs_client/test/gcs_client_test.cc (hunk 9 above)
Comment (iycheng) [reply src/ray/gcs/gcs_client/test/gcs_client_test.cc:907]: I think in the end, we need to test with two groups of setups for different storage.

File: python/ray/_private/services.py [hunk 12]
```diff
@@ -1778,14 +1779,14 @@ def build_java_worker_command(
...
-def build_cpp_worker_command(cpp_worker_options, redis_address,
+def build_cpp_worker_command(cpp_worker_options, bootstrap_address,
                              plasma_store_name, raylet_name, redis_password,
                              session_dir, log_dir, node_ip_address):
     """This method assembles the command used to start a CPP worker.
 
     Args:
         cpp_worker_options (list): The command options for CPP worker.
-        redis_address (str): Redis address of GCS.
+        bootstrap_address (str): The bootstrap of the cluster.
```
Comment (mwtian) [inline python/ray/_private/services.py:1789]: The bootstrap address?

**Review by mwtian — APPROVED** [review]

**Comment by mwtian** [comment]
Looks good!

**Pull request merged**